File: optimization/tests/test_tools.py
# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at
# sgpp.sparsegrids.org

import unittest
import pysgpp
import random
import tempfile

class TestTools(unittest.TestCase):
    def setUp(self):
        """Initialize the test case."""
        # disable status output
        pysgpp.cvar.OptPrinterInstance.setVerbosity(-1)
        # disable multi-threading
        pysgpp.omp_set_num_threads(1)
    
    def gridEqualityTest(self, grid1, grid2):
        grid1Storage = grid1.getStorage()
        grid2Storage = grid2.getStorage()
        d = grid1Storage.dim()
        self.assertEqual(grid1Storage.size(), grid2Storage.size())
        self.assertEqual(d, grid2Storage.dim())
        
        for k in range(grid1Storage.size()):
            for t in range(d):
                self.assertEqual(grid1Storage.get(k).getLevel(t),
                                 grid2Storage.get(k).getLevel(t))
                self.assertEqual(grid1Storage.get(k).getIndex(t),
                                 grid2Storage.get(k).getIndex(t))
    
    def testFileIOReadWriteGrid(self):
        """Test SGPP::optimization::file_io::readGrid/writeGrid."""
        random.seed(42)
        
        for d in range(1, 6):
            grid1 = pysgpp.Grid.createLinearGrid(d)
            gridGen = grid1.createGridGenerator()
            gridGen.regular(3)
            
            with tempfile.NamedTemporaryFile() as f:
                pysgpp.OptFileIOWriteGrid(f.name, grid1.getStorage())
                grid2 = pysgpp.Grid.createLinearGrid(d)
                pysgpp.OptFileIOReadGrid(f.name, grid2.getStorage())
            
            self.gridEqualityTest(grid1, grid2)
            
            functionValues1 = pysgpp.DataVector(
                    [random.uniform(0.0, 1.0)
                     for k in range(grid1.getStorage().size())])
            
            with tempfile.NamedTemporaryFile() as f:
                pysgpp.OptFileIOWriteGrid(f.name, grid1.getStorage(),
                                          functionValues1)
                grid2 = pysgpp.Grid.createLinearGrid(d)
                functionValues2 = pysgpp.DataVector(0)
                pysgpp.OptFileIOReadGrid(f.name, grid2.getStorage(),
                                         functionValues2)
            
            self.gridEqualityTest(grid1, grid2)
            self.assertEqual(len(functionValues1), len(functionValues2))
            for k in range(len(functionValues1)):
                if pysgpp.cvar.USING_DOUBLE_PRECISION:
                    self.assertEqual(functionValues1[k], functionValues2[k])
                else:
                    self.assertAlmostEqual(functionValues1[k], functionValues2[k])
    
    def testFileIOReadWriteMatrix(self):
        """Test SGPP::optimization::file_io::readMatrix/writeMatrix."""
        random.seed(42)
        m1, n1 = (100, 200)
        
        # test read/write with std::vector<SGPP::float_t>
        A1 = pysgpp.DoubleVector()
        for i in range(m1):
            for j in range(n1):
                A1.push_back(random.uniform(0.0, 1.0))
        
        with tempfile.NamedTemporaryFile() as f:
            pysgpp.OptFileIOWriteMatrix(f.name, A1, m1, n1)
            A2 = pysgpp.DoubleVector()
            m2, n2 = pysgpp.OptFileIOReadMatrix(f.name, A2)
        
        self.assertEqual(m1, m2)
        self.assertEqual(n1, n2)
        self.assertEqual(len(A1), len(A2))
        for k in range(len(A1)):
            self.assertEqual(A1[k], A2[k])
        
        # test read/write with DataMatrix
        A1 = pysgpp.DataMatrix(m1, n1)
        for i in range(m1):
            for j in range(n1):
                A1.set(i, j, random.uniform(0.0, 1.0))
        
        with tempfile.NamedTemporaryFile() as f:
            pysgpp.OptFileIOWriteMatrix(f.name, A1)
            A2 = pysgpp.DataMatrix(0, 0)
            pysgpp.OptFileIOReadMatrix(f.name, A2)
        
        self.assertEqual(m1, m2)
        self.assertEqual(n1, n2)
        self.assertEqual(A1.getNrows(), A2.getNrows())
        self.assertEqual(A1.getNcols(), A2.getNcols())
        for i in range(m1):
            for j in range(n1):
                self.assertEqual(A1.get(i, j), A2.get(i, j))
    
    def testFileIOReadWriteVector(self):
        """Test SGPP::optimization::file_io::readVector/writeVector."""
        random.seed(42)
        n = 100
        
        # test read/write with std::vector<SGPP::float_t>
        v1 = pysgpp.DoubleVector()
        for i in range(n):
            v1.push_back(random.uniform(0.0, 1.0))
        
        with tempfile.NamedTemporaryFile() as f:
            pysgpp.OptFileIOWriteVector(f.name, v1)
            v2 = pysgpp.DoubleVector()
            pysgpp.OptFileIOReadVector(f.name, v2)
        
        self.assertEqual(len(v1), len(v2))
        for k in range(len(v1)):
            self.assertEqual(v1[k], v2[k])
        
        # test read/write with DataVector
        v1 = pysgpp.DataVector(n)
        for i in range(n):
            v1[i] = random.uniform(0.0, 1.0)
        
        with tempfile.NamedTemporaryFile() as f:
            pysgpp.OptFileIOWriteVector(f.name, v1)
            v2 = pysgpp.DataVector(0)
            pysgpp.OptFileIOReadVector(f.name, v2)
        
        self.assertEqual(len(v1), len(v2))
        for i in range(n):
            self.assertEqual(v1[i], v2[i])
    
    def testRandomNumberGenerator(self):
        """Test SGPP::optimization::RandomNumberGenerator."""
        seed = 42
        N = 10000
        
        # set and test seed getting/setting
        pysgpp.cvar.OptRNGInstance.setSeed()
        pysgpp.cvar.OptRNGInstance.setSeed(seed)
        self.assertEqual(pysgpp.cvar.OptRNGInstance.getSeed(), seed)
        
        # test continuous uniform random numbers
        numbers = [pysgpp.cvar.OptRNGInstance.getUniformRN()
                   for i in range(N)]
        mean = sum(numbers) / len(numbers)
        var = sum([(number - mean)**2 for number in numbers]) / len(numbers)
        
        self.assertTrue(all([isinstance(number, float) for number in numbers]))
        self.assertTrue(all([0.0 <= number <= 1.0 for number in numbers]))
        self.assertAlmostEqual(mean, 0.5, places=2)
        self.assertAlmostEqual(var, 1.0/12.0, places=2)
        
        # test Gaussian random numbers
        mus = [0.0, 12.3, -42.0, 13.37]
        sigmas = [1.0, 2.6, 8.1, 0.3]
        
        for mu, sigma in zip(mus, sigmas):
            numbers = [pysgpp.cvar.OptRNGInstance.getGaussianRN(sigma, mu)
                       for i in range(N)]
            mean = sum(numbers) / len(numbers)
            var = sum([(number - mean)**2 for number in numbers]) / \
                  len(numbers)
                  
            self.assertTrue(all([isinstance(number, float)
                                 for number in numbers]))
            self.assertAlmostEqual(mean, mu, delta=sigma*0.1)
            self.assertAlmostEqual(var, sigma*sigma, delta=sigma*sigma*0.1)
        
        # test discrete uniform random numbers
        for k in range(1, 11):
            numbers = [pysgpp.cvar.OptRNGInstance.getUniformIndexRN(k)
                       for i in range(N)]
            mean = float(sum(numbers)) / float(len(numbers))
            var = float(sum([(number - mean)**2 for number in numbers])) / \
                  float(len(numbers))
            
            self.assertTrue(all([isinstance(number, int) or
                                 isinstance(number, long)
                                 for number in numbers]))
            self.assertTrue(all([0 <= number <= k-1 for number in numbers]))
            self.assertAlmostEqual(mean, (k-1.0)/2.0, delta=k*0.01)
            self.assertAlmostEqual(var, (k*k-1.0)/12.0, delta=k*k*0.01)
    
    def orthogonalityTest(self, A):
        n = A.getNrows()
        self.assertEqual(n, A.getNcols())
        
        for i in range(n):
            for j in range(i+1):
                entry = 0.0
                for l in range(n):
                    entry += A.get(i, l) * A.get(j, l)
                self.assertAlmostEqual(entry, 1.0 if (i == j) else 0.0)
    
    def symmetryTest(self, A):
        n = A.getNrows()
        self.assertEqual(n, A.getNcols())
        
        for i in range(n):
            for j in range(i):
                self.assertAlmostEqual(A.get(i, j), A.get(j, i))
    
    def similiarityTest(self, A, V, B):
        n = A.getNrows()
        self.assertEqual(n, A.getNcols())
        self.assertEqual(n, V.getNrows())
        self.assertEqual(n, V.getNcols())
        self.assertEqual(n, B.getNrows())
        self.assertEqual(n, B.getNcols())
        
        for i in range(n):
            for j in range(n):
                entry1 = 0.0
                entry2 = 0.0
                for l in range(n):
                    entry1 += A.get(i, l) * V.get(l, j)
                    entry2 += V.get(i, l) * B.get(l, j)
                self.assertAlmostEqual(entry1, entry2)
    
    def generateRandomMatrix(self, n):
        A = pysgpp.DataMatrix(n, n)
        for i in range(n):
            for j in range(n):
                A.set(i, j, random.gauss(0.0, 1.0))
        return A
    
    def testMathHouseholderTransformation(self):
        """Test SGPP::optimization::math::householderTransformation."""
        random.seed(42)
        n = 20
        A = self.generateRandomMatrix(n)
        
        for p, q in [(0, 0), (2, 1), (5, 9)]:
            m = n - p
            Q = pysgpp.DataMatrix(m, m)
            pysgpp.OptMathHouseholderTransformation(A, p, q, Q)
            self.symmetryTest(Q)
            self.orthogonalityTest(Q)
            
            for i in range(p+1, n):
                entry = 0.0
                for l in range(p, n):
                    entry += Q.get(i-p, l-p) * A.get(l, q)
                self.assertAlmostEqual(entry, 0.0)
    
    def testMathHessenbergForm(self):
        """Test SGPP::optimization::math::hessenbergForm."""
        random.seed(42)
        n = 20
        A = self.generateRandomMatrix(n)
        H = pysgpp.DataMatrix(A)
        V = pysgpp.DataMatrix(n, n)
        pysgpp.OptMathHessenbergForm(H, V)
        self.orthogonalityTest(V)
        self.similiarityTest(A, V, H)
        
        for i in range(n):
            for j in range(i-2):
                self.assertAlmostEqual(H.get(i, j), 0.0)
    
    def testMathQRDecomposition(self):
        """Test SGPP::optimization::math::QRDecomposition."""
        random.seed(42)
        n = 20
        A = self.generateRandomMatrix(n)
        R = pysgpp.DataMatrix(A)
        Q = pysgpp.DataMatrix(n, n)
        pysgpp.OptMathQRDecomposition(R, Q)
        self.orthogonalityTest(Q)
        
        for i in range(n):
            for j in range(i-1):
                self.assertAlmostEqual(R.get(i, j), 0.0)
        
        for i in range(n):
            for j in range(n):
                entry = 0.0
                for l in range(n):
                    entry += Q.get(i, l) * R.get(l, j)
                self.assertAlmostEqual(A.get(i, j), entry)
    
    def testMathSchurDecomposition(self):
        """Test SGPP::optimization::math::schurDecomposition."""
        random.seed(42)
        n = 20
        A = self.generateRandomMatrix(n)
        
        for i in range(n):
            for j in range(i):
                A.set(i, j, A.get(j, i))
        
        S = pysgpp.DataMatrix(A)
        V = pysgpp.DataMatrix(n, n)
        pysgpp.OptMathSchurDecomposition(S, V)
        self.orthogonalityTest(V)
        self.similiarityTest(A, V, S)
        
        for i in range(n):
            for j in range(n):
                if i != j:
                    self.assertAlmostEqual(S.get(i, j), 0.0)

File: optimization/tests/testTools.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>

#include <sgpp/base/grid/Grid.hpp>
#include <sgpp/base/tools/Printer.hpp>
#include <sgpp/base/tools/RandomNumberGenerator.hpp>
#include <sgpp/base/tools/sle/system/FullSLE.hpp>
#include <sgpp/optimization/gridgen/IterativeGridGeneratorRitterNovak.hpp>
#include <sgpp/optimization/test_problems/unconstrained/Sphere.hpp>
#include <sgpp/optimization/tools/FileIO.hpp>
#include <sgpp/optimization/tools/Math.hpp>

#include <string>
#include <vector>

#include "ObjectiveFunctions.hpp"

using sgpp::base::Printer;
using sgpp::base::RandomNumberGenerator;

void gridEqualityTest(sgpp::base::Grid& grid1, sgpp::base::Grid& grid2) {
  sgpp::base::GridStorage& storage1 = grid1.getStorage();
  sgpp::base::GridStorage& storage2 = grid2.getStorage();
  const size_t d = storage1.getDimension();
  BOOST_CHECK_EQUAL(d, storage2.getDimension());
  const size_t n = storage1.getSize();
  BOOST_CHECK_EQUAL(n, storage2.getSize());

  for (size_t k = 0; k < n; k++) {
    for (size_t t = 0; t < d; t++) {
      BOOST_CHECK_EQUAL(storage1[k].getLevel(t), storage2[k].getLevel(t));
      BOOST_CHECK_EQUAL(storage1[k].getIndex(t), storage2[k].getIndex(t));
    }
  }
}

void orthogonalityTest(sgpp::base::DataMatrix& A) {
  const size_t n = A.getNrows();
  BOOST_CHECK_EQUAL(n, A.getNcols());

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j <= i; j++) {
      double entry = 0.0;

      for (size_t l = 0; l < n; l++) {
        entry += A(i, l) * A(j, l);
      }

      BOOST_CHECK_SMALL(entry - ((i == j) ? 1.0 : 0.0), 1e-10);
    }
  }
}

void symmetryTest(sgpp::base::DataMatrix& A) {
  const size_t n = A.getNrows();
  BOOST_CHECK_EQUAL(n, A.getNcols());

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < i; j++) {
      BOOST_CHECK_CLOSE(A(i, j), A(j, i), 1e-10);
    }
  }
}

void similiarityTest(sgpp::base::DataMatrix& A, sgpp::base::DataMatrix& V,
                     sgpp::base::DataMatrix& B) {
  const size_t n = A.getNrows();
  BOOST_CHECK_EQUAL(n, A.getNcols());
  BOOST_CHECK_EQUAL(n, V.getNrows());
  BOOST_CHECK_EQUAL(n, V.getNcols());
  BOOST_CHECK_EQUAL(n, B.getNrows());
  BOOST_CHECK_EQUAL(n, B.getNcols());

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < n; j++) {
      double entry1 = 0.0, entry2 = 0.0;

      for (size_t l = 0; l < n; l++) {
        entry1 += A(i, l) * V(l, j);
        entry2 += V(i, l) * B(l, j);
      }

      BOOST_CHECK_CLOSE(entry1, entry2, 5e-5);
    }
  }
}

void generateRandomMatrix(sgpp::base::DataMatrix& A) {
  for (size_t i = 0; i < A.getNrows(); i++) {
    for (size_t j = 0; j < A.getNcols(); j++) {
      A(i, j) = RandomNumberGenerator::getInstance().getGaussianRN();
    }
  }
}

template <class T>
void randomMatrixEntry(T& x) {
  x = static_cast<T>(RandomNumberGenerator::getInstance().getUniformIndexRN(100));
}

template <>
void randomMatrixEntry(float& x) {
  x = static_cast<float>(RandomNumberGenerator::getInstance().getUniformRN(-100.0, 100.0));
}

template <>
void randomMatrixEntry(double& x) {
  x = static_cast<double>(RandomNumberGenerator::getInstance().getUniformRN(-100.0, 100.0));
}

template <>
void randomMatrixEntry(std::string& x) {
  const size_t length = RandomNumberGenerator::getInstance().getUniformIndexRN(100);
  x.clear();

  for (size_t i = 0; i < length; i++) {
    x += static_cast<char>(32 + RandomNumberGenerator::getInstance().getUniformIndexRN(96));
  }
}

template <class T>
void testReadWriteMatrix(std::vector<T>& A1, size_t m1, size_t n1) {
  std::vector<T> A2;
  A1.clear();

  for (size_t i = 0; i < m1; i++) {
    for (size_t j = 0; j < n1; j++) {
      T entry;
      randomMatrixEntry(entry);
      A1.push_back(entry);
    }
  }

  size_t m2, n2;

  {
    const std::string fileName = "testTools_matrix.tmp";
    sgpp::optimization::file_io::writeMatrix(fileName, A1, m1, n1);
    sgpp::optimization::file_io::readMatrix(fileName, A2, m2, n2);
    std::remove(fileName.c_str());
  }

  BOOST_CHECK_EQUAL(m1, m2);
  BOOST_CHECK_EQUAL(n1, n2);
  BOOST_CHECK_EQUAL(A1.size(), A2.size());

  for (size_t k = 0; k < A1.size(); k++) {
    BOOST_CHECK_EQUAL(A1[k], A2[k]);
  }
}

BOOST_AUTO_TEST_CASE(TestFileIOReadWriteGrid) {
  // Test sgpp::optimization::sgpp::optimization::file_io::readGrid/writeGrid.
  Printer::getInstance().setVerbosity(-1);
  RandomNumberGenerator::getInstance().setSeed(42);

  for (size_t d = 1; d < 6; d++) {
    std::unique_ptr<sgpp::base::Grid> grid1(sgpp::base::Grid::createLinearGrid(d)),
        grid2(sgpp::base::Grid::createLinearGrid(d));
    grid1->getGenerator().regular(3);

    {
      const std::string fileName = "testTools_grid.tmp";
      sgpp::optimization::file_io::writeGrid(fileName, grid1->getStorage());
      sgpp::optimization::file_io::readGrid(fileName, grid2->getStorage());
      std::remove(fileName.c_str());
    }

    gridEqualityTest(*grid1, *grid2);

    sgpp::base::DataVector functionValues1(grid1->getSize());
    sgpp::base::DataVector functionValues2(0);

    for (size_t k = 0; k < functionValues1.getSize(); k++) {
      functionValues1[k] = RandomNumberGenerator::getInstance().getUniformRN();
    }

    {
      const std::string fileName = "testTools_grid.tmp";
      sgpp::optimization::file_io::writeGrid(fileName, grid1->getStorage(), functionValues1);
      std::unique_ptr<sgpp::base::Grid> grid2(sgpp::base::Grid::createLinearGrid(d));
      sgpp::optimization::file_io::readGrid(fileName, grid2->getStorage(), functionValues2);
      std::remove(fileName.c_str());
    }

    gridEqualityTest(*grid1, *grid2);
    BOOST_CHECK_EQUAL(functionValues1.getSize(), functionValues2.getSize());

    for (size_t k = 0; k < functionValues1.getSize(); k++) {
      BOOST_CHECK_EQUAL(functionValues1[k], functionValues2[k]);
    }
  }
}

BOOST_AUTO_TEST_CASE(TestFileIOReadWriteMatrix) {
  // Test sgpp::optimization::sgpp::optimization::file_io::readMatrix/writeMatrix.
  Printer::getInstance().setVerbosity(-1);
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t m1 = 100;
  const size_t n1 = 200;

  // test read/write with std::vector<T>
  {
    std::vector<double> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<float> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<double> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<uint8_t> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<uint16_t> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<uint32_t> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<uint64_t> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  {
    std::vector<std::string> A1;
    testReadWriteMatrix(A1, m1, n1);
  }

  // test read/write with DataMatrix
  {
    sgpp::base::DataMatrix A1(m1, n1), A2(0, 0);

    for (size_t i = 0; i < m1; i++) {
      for (size_t j = 0; j < n1; j++) {
        A1(i, j) = RandomNumberGenerator::getInstance().getUniformRN();
      }
    }

    {
      const std::string fileName = "testTools_matrix.tmp";
      sgpp::optimization::file_io::writeMatrix(fileName, A1);
      sgpp::optimization::file_io::readMatrix(fileName, A2);
      std::remove(fileName.c_str());
    }

    BOOST_CHECK_EQUAL(A1.getNrows(), A2.getNrows());
    BOOST_CHECK_EQUAL(A1.getNcols(), A2.getNcols());

    for (size_t i = 0; i < m1; i++) {
      for (size_t j = 0; j < n1; j++) {
        BOOST_CHECK_EQUAL(A1(i, j), A2(i, j));
      }
    }
  }
}

BOOST_AUTO_TEST_CASE(TestFileIOReadWriteVector) {
  // Test sgpp::optimization::sgpp::optimization::file_io::readVector/writeVector.
  Printer::getInstance().setVerbosity(-1);
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t n = 100;

  // test read/write with std::vector<double>
  {
    std::vector<double> v1, v2;

    for (size_t i = 0; i < n; i++) {
      v1.push_back(RandomNumberGenerator::getInstance().getUniformRN());
    }

    {
      const std::string fileName = "testTools_matrix.tmp";
      sgpp::optimization::file_io::writeVector(fileName, v1);
      sgpp::optimization::file_io::readVector(fileName, v2);
      std::remove(fileName.c_str());
    }

    BOOST_CHECK_EQUAL(v1.size(), v2.size());

    for (size_t k = 0; k < v1.size(); k++) {
      BOOST_CHECK_EQUAL(v1[k], v2[k]);
    }
  }

  // test read/write with DataVector
  {
    sgpp::base::DataVector v1(n), v2(0);

    for (size_t i = 0; i < n; i++) {
      v1[i] = RandomNumberGenerator::getInstance().getUniformRN();
    }

    {
      const std::string fileName = "testTools_vector.tmp";
      sgpp::optimization::file_io::writeVector(fileName, v1);
      sgpp::optimization::file_io::readVector(fileName, v2);
      std::remove(fileName.c_str());
    }

    BOOST_CHECK_EQUAL(v1.getSize(), v2.getSize());

    for (size_t i = 0; i < v1.getSize(); i++) {
      BOOST_CHECK_EQUAL(v1[i], v2[i]);
    }
  }
}

BOOST_AUTO_TEST_CASE(TestHouseholderTransformation) {
  // Test sgpp::optimization::sgpp::optimization::math::householderTransformation.
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t n = 20;
  sgpp::base::DataMatrix A(n, n);
  generateRandomMatrix(A);

  const std::vector<size_t> ps = {0, 2, 5};
  const std::vector<size_t> qs = {0, 1, 9};

  for (size_t k = 0; k < ps.size(); k++) {
    const size_t p = ps[k];
    const size_t q = qs[k];
    const size_t m = n - p;
    sgpp::base::DataMatrix Q(m, m);
    sgpp::optimization::math::householderTransformation(A, p, q, Q);
    symmetryTest(Q);
    orthogonalityTest(Q);

    for (size_t i = p + 1; i < n; i++) {
      double entry = 0.0;

      for (size_t l = p; l < n; l++) {
        entry += Q(i - p, l - p) * A(l, q);
      }

      BOOST_CHECK_SMALL(entry, 1e-10);
    }
  }
}

BOOST_AUTO_TEST_CASE(TestHessenbergForm) {
  // Test sgpp::optimization::sgpp::optimization::math::hessenbergForm.
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t n = 20;
  sgpp::base::DataMatrix A(n, n);
  generateRandomMatrix(A);
  sgpp::base::DataMatrix H(A);
  sgpp::base::DataMatrix V(n, n);
  sgpp::optimization::math::hessenbergForm(H, V);

  orthogonalityTest(V);
  similiarityTest(A, V, H);

  for (size_t i = 2; i < n; i++) {
    for (size_t j = 0; j < i - 1; j++) {
      BOOST_CHECK_SMALL(H(i, j), 1e-10);
    }
  }
}

BOOST_AUTO_TEST_CASE(TestQRDecomposition) {
  // Test sgpp::optimization::sgpp::optimization::math::QRDecomposition.
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t n = 20;
  sgpp::base::DataMatrix A(n, n);
  generateRandomMatrix(A);
  sgpp::base::DataMatrix R(A);
  sgpp::base::DataMatrix Q(n, n);
  sgpp::optimization::math::QRDecomposition(R, Q);

  orthogonalityTest(Q);

  for (size_t i = 1; i < n; i++) {
    for (size_t j = 0; j < i - 1; j++) {
      BOOST_CHECK_SMALL(R(i, j), 1e-10);
    }
  }

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < n; j++) {
      double entry = 0.0;

      for (size_t l = 0; l < n; l++) {
        entry += Q(i, l) * R(l, j);
      }

      BOOST_CHECK_CLOSE(A(i, j), entry, 1e-10);
    }
  }
}

BOOST_AUTO_TEST_CASE(TestSchurDecomposition) {
  // Test sgpp::optimization::sgpp::optimization::math::schurDecomposition.
  RandomNumberGenerator::getInstance().setSeed(42);
  const size_t n = 20;
  sgpp::base::DataMatrix A(n, n);
  generateRandomMatrix(A);

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < i; j++) {
      A(i, j) = A(j, i);
    }
  }

  sgpp::base::DataMatrix S(A);
  sgpp::base::DataMatrix V(n, n);
  sgpp::optimization::math::schurDecomposition(S, V);

  orthogonalityTest(V);
  similiarityTest(A, V, S);

  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < n; j++) {
      if (i != j) {
        BOOST_CHECK_SMALL(S(i, j), 1e-8);
      }
    }
  }
}
